python/lsst/cbp/coordUtils.py

```diff
@@ -0,0 +1,400 @@
+
+    telBase = SpherePoint(0, 0, radians)
+
+    # rotate vector around base z axis by -daz
+    daz = pupilAzAlt[0] - telBase[0]
+    zaxis = SpherePoint(Vector3d(0, 0, 1))
+    vectorSpRot1 = vectorSpBase.rotated(axis=zaxis, amount=-daz)
+
+    # rotate vector around pupil -y axis by -dalt
+    dalt = pupilAzAlt[1] - telBase[1]
+    negYAxis = SpherePoint(Vector3d(0, -1, 0))
+    vectorSpPupil = vectorSpRot1.rotated(axis=negYAxis, amount=-dalt)
+    return np.array(vectorSpPupil.getVector()) * vectorMag
+
+
+def convertVectorFromPupilToBase(vectorPupil, pupilAzAlt):
+    """Given a vector in pupil coordinates and the pupil pointing,
+    compute the vector in base coords
+
+    Parameters
+    ----------
+    vectorPupil: sequence of 3 `float`
+        3-dimesional vector in the
+        :ref:`pupil frame <lsst.cbp.pupil_frame>`
+    pupilAzAlt: `lsst.afw.geom.SpherePoint`
+        Pointing of the pupil frame as
+        :ref:`internal azimuth, altitude <lsst.cbp.internal_angles>`
+
+    Returns
+    -------
+    vectorBase: `np.array` of 3 `float`
+        3-dimensional vector in the :ref:`base frame <lsst.cbp.base_frame>`
+    """
+    vectorMag = np.linalg.norm(vectorPupil)
+    vectorSpPupil = SpherePoint(Vector3d(*vectorPupil))
+
+    telBase = SpherePoint(0, 0, radians)
+
+    # rotate vector around pupil -y axis by dalt
+    dalt = pupilAzAlt[1] - telBase[1]
+    negYAxis = SpherePoint(Vector3d(0, -1, 0))
+    vectorSpRot1 = vectorSpPupil.rotated(axis=negYAxis, amount=dalt)
+
+    # rotate that around base z axis by daz
+    daz = pupilAzAlt[0] - telBase[0]
+    zaxis = SpherePoint(Vector3d(0, 0, 1))
+    vectorSpBase = vectorSpRot1.rotated(axis=zaxis, amount=daz)
+    return np.array(vectorSpBase.getVector()) * vectorMag
+
+
+def computeAzAltFromBasePupil(vectorBase, vectorPupil):
+    """Compute az/alt from a vector to in the base frame
+    and the same vector in the pupil frame
+
+    Parameters
+    ----------
+    vectorBase : `iterable` of three `float`
+        3-dimensional vector in the
+        :ref:`base frame <lsst.cbp.base_frame>`
+    vectorPupil : `iterable` of `float`
+        The same vector in the :ref:`pupil frame <lsst.cbp.pupil_frame>`.
+        This vector should be within 45 degrees or so of the optical axis
+        for accurate results.
+
+    Returns
+    -------
+    pupilAzAlt : `lsst.afw.geom.SpherePoint`
+        Pointing of the pupil frame as
+        :ref:`internal azimuth, altitude <lsst.cbp.internal_angles>`
+
+    Notes
+    -----
+    The magnitude of each vector is ignored, except that a reasonable
+    magnitude is required in order to compute an accurate unit vector.
+    """
+    spBase = SpherePoint(Vector3d(*vectorBase))
+    spPupil = SpherePoint(Vector3d(*vectorPupil))
+
+    # compute telescope altitude
+    # Use an equation derived from the following, which is from Euler angles:
+    # base z = -sin(negAlt) pupil x + cos(negAlt) pupil z
+    #        = sin(alt) pupil x + cos(alt) pupil z
```

Comment:
Invoking the Euler angles was a bit misleading for me, as this system only has two angles (alt, az) instead of three and the equation for base z (given above in your comment) can be derived by using simple 2D trig once you know that base z is always in the pupil x-z plane. When `xp=0` or `zp=0` you can use that equation to solve for `zb` directly. Otherwise you have 4 quadrants to worry about based on the values of `xp` and `zp`, and it turns out that (unless I made a sign error), you should be able to replace lines 324 to 342 with:

```
if xp == 0:
    alt = math.acos(zb/zp)
elif zp == 0:
    alt = math.asin(zb/xp)
else:
    if xp > 0:
        sinAlt = factor * (addend1 + addend2)
    else:
        sinAlt = factor * (addend1 - addend2)
    alt = math.asin(sinAlt)*radians
```

Reply:
Aside from a few missing `*radians` and a sign error in applying addend2 this works and is significantly simpler than the original code. Thanks!

As to the Euler angles...that's a very good point. I have included a fully detailed description of the Euler matrix in the doc string for convertVectorFromBaseToPupil and reference the last row of that matrix as justification for `base z  = sin(alt) pupil x + cos(alt) pupil z`